Tools: Add support for pre-commit

projectgus opened the pull request:
Adds documented support for [pre-commit](https://pre-commit.com/) as an easy way to run the CI code formatting & commit message checks locally.

I went the long way around here, originally trying to adapt existing pre-commit plugins to the MicroPython CI requirements but in the end it was easier to tweak `tools/codeformat.py` and `tools/verifygitlog.py` so they could be easily run from pre-commit.

Also add some documentation about the supported uncrustify versions, as it turns out they are not all the same.

## Extra Information

<details>
<summary>Some notes about differences between existing pre-commit hooks and the MicroPython CI checks here.</summary>


These notes are only for future reference in case anyone wants to try again to use hooks in the future. It's unclear what the benefit would be, unless there are other useful pre-commit hooks added then it's useful to run pre-commit both in CI and locally and get the exact same results.

**uncrustify** has some additional fixups applied by the `fixup_c` function in codeformat.py that will be hard to recreate. Maybe astyle or another tool would be able to approximate them, though?

**black** pretty much works out of the box, some work to adapt file exclusions & inclusions to a multi-line regex, i.e.
```yaml
  - repo: https://github.com/psf/black
    rev: 22.6.0
    hooks:
      - id: black
        args: [ --fast, --line-length=99 ]
        # If adding exclusions here, also add them in tools/codeformat.py
        exclude: |
            (?x)^(
                tests/.+/repl_.+|
                tests/basics/.+
            )$
```
(This currently applies to all .py files, whereas codeformat.py only looks in some directories.)

[gitlint](https://jorisroovers.com/gitlint/) works but is a bit fiddly to configure some of the message requirements exactly the same as verifygitlog.py, for example allowing long lines that contain URLs. A work in progress .gitlint file is:

```
[general]
contrib=contrib-body-requires-signed-off-by

[title-max-length]
line-length=72

[body-max-line-length]
line-length=75

[title-match-regex]
regex=^[^!]+: [A-Z]+.+ .+\.$

[author-valid-email]
regex=(?!noreply)
```
</details>

*This work was funded through GitHub Sponsors.*

projectgus committed:
tools: Add note about uncrustify versions.

Uncrustify versions are not mutually compatible:

1. Version 0.73 or newer produce slightly different formatting. It may be
possible to tweak these by adding more config items, but this will cause
older versions to error out with 'Unknown option'.

2. Version 0.75 prints a range of deprecation warnings due to config file
changes, and returns a non-zero exit code. These are actually fixable
as most are the default value, and pp_indent has changed from 'true' to '1'
which is backwards compatible. However issue 1 remains, so probably better
to have it fail explicitly.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
tools: Add pre-commit support.

Tweak the existing codeformat.py and verifygitlog.py to allow them to be
easily called by pre-commit.

(This turned out to be easier than using any existing pre-commit hooks,
without making subtle changes in the formatting.)

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge @jimmo PTAL, I think this may be useful to document for new contributors (and I've found it useful for myself!)

projectgus commented:
OK, that's somewhat amusing that "black" is failing on the commit that adds the pre-commit checks...

Obviously it passes here, so I think this is a "black" version difference between my local version and the version used in CI. This is one of the things that pre-commit can help with, so I'll update the PR again soon to fix it.

stinos commented:
There are a couple of pre-commit frameworks, could you explain why this one was chosen?

jimmo commented:
Thanks @projectgus -- I looked into this a while back but ran into several issues and ended up applying https://xkcd.com/1205/   I don't think I came across `pre-commit` though so that might have saved me a lot of effort. Looking at the docs, I think `pre-commit` will likely solve all the issues I came across.

One thing I'm confused by though, how does pre-commit know to run codeformat.py? Is there a `.pre-commit-config.yaml` that should be added to this PR?

> uncrustify has some additional fixups applied by the fixup_c function in codeformat.py that will be hard to recreate. Maybe astyle or another tool would be able to approximate them, though?

Some light reading for anyone interested in the history: #4223 #5631 #5632 #5668 #5671 #5691 (I think there might have been more). In summary, we tried clang-format, astyle, and uncrustify. uncrustify got the closest. It's possible that either of these tools has subsequently gained the feature we need (mostly about indentation of preprocessor).

projectgus commented:
> Thanks @projectgus -- I looked into this a while back but ran into several issues and ended up applying https://xkcd.com/1205/ 

Haha fair enough! For me, I find bouncing off CI mildly demoralising and it seems to happen to me on every PR for something relatively minor. (I'm also not disciplined enough to always manually run the checks before pushing.) So the benefit from my perspective is partially time and partially emotional wellbeing.

If you remember any particular concerns then feel free to mention them, can double check they won't come to bite us here.

> One thing I'm confused by though, how does pre-commit know to run codeformat.py? Is there a .pre-commit-config.yaml that should be added to this PR?

:facepalm: I must have unstaged it during some last minute commit cleanup. Sorry, fixed now.

> There are a couple of pre-commit frameworks, could you explain why this one was chosen?

@stinos Very reasonable question!

The first thing to acknowledge is that pre-commit requirements for MicroPython, at least as implemented here, are really minor. So why use one at all instead of publishing steps to edit the git config file?

* Simpler steps, less friction for new contributors.
* The hook is fully defined by a single file versioned in the repo, so if it ever needs to change in the future there's minimal risk of breaking existing hooks.

These requirements can likely be met by *any* pre-commit hook framework, so why pick [pre-commit](https://pre-commit.com/)?

* Written in pure Python, so it should install directly into whatever environment a MicroPython developer already has.
* Mature, widely supported, widely understood.
* I was familiar with it. :)

There are a bunch of other useful features that turned out not to be used here:

* Good official support for both Python ecosystem tools (i.e. [black](https://black.readthedocs.io/en/stable/integrations/source_version_control.html)) and tools written in Python (i.e. [gitlint](https://jorisroovers.com/gitlint/#using-gitlint-through-pre-commit)), without being limited to only using Python.
* Versioned install environments to automatically install and use a pinned version of a particular tool. Useful in the very unlikely case of a black behaviour change. The real useful application for this would be managing a pinned "uncrustify" version automatically, but this isn't supported out of the box (uncrustify is a C++ program). If there's a pre-commit framework that can do this out of the box then we should use that!

Theose features may come in useful in the future if there's another linting tool or similar that's desirable to add, but they quite likely may not.

> Some light reading for anyone interested in the history: https://github.com/micropython/micropython/issues/4223 https://github.com/micropython/micropython/pull/5631 https://github.com/micropython/micropython/pull/5632 https://github.com/micropython/micropython/pull/5668 https://github.com/micropython/micropython/pull/5671 https://github.com/micropython/micropython/pull/5691 (I think there might have been more). In summary, we tried clang-format, astyle, and uncrustify. uncrustify got the closest. It's possible that either of these tools has subsequently gained the feature we need (mostly about indentation of preprocessor).

Right, sorry @jimmo that was thoughtless of me to assume you hadn't already carefully looked into this and I should have dug deeper into the history first!

I think the uncrustify+fixup_c() approach is absolutely fine.

stinos commented on CODECONVENTIONS.md:
```diff
@@ -65,6 +65,31 @@ changes to the correct style.  Without arguments this tool will reformat all
+Automatic Pre-Commit Hooks
+==========================
+
+To have commits automatically checked using
```
Perhaps explain briefly what 'checked' entails?

projectgus replied:
Great idea, will do!

projectgus replied:
Have added some explanation plus some tips

stinos commented on tools/codeformat.py:
```diff
@@ -151,6 +151,11 @@ def main():
+    cmd_parser.add_argument(
+        "-f",
+        action="store_true",
+        help="Treat files argument as a list of names not globbing patterns, and apply default inclusion/exclusion filters",
```
Just wondering: is this neeed? Expanding a glob which has no globbing specifier in it just yields the original string already? I mean, doesn't codeformat.py already work like this so you just pass complete paths and it does what is needed?

projectgus replied:
It's needed to apply the include/exclude filter. If you do `codeformat.py somefile.c` then that file gets checked always - even if is not included the PATHS list, or if it's excluded in the EXCLUSIONS list (i.e. not checked in CI).

However for the pre-commit hook we only want to check a file if it would also be checked in CI.

I'll rename the help text to make it clearer, something like "If a list of files is given on the command line, filter the list against paths to include/exclude from the default CI check."

stinos replied:
Ok sorry I didn't read the rest of the code and wrongfully assumed it would always apply filtering, also because that seems to be the common way formatting tools work

projectgus replied:
Yeah, no problems it's a bit fiddly. It is useful to bypass the filters sometimes, i.e. if you want to check if a file or directory can be added to the CI check. I've rewritten the help text a bit to try and make it clearer.

stinos replied:
For consistency I'd reword that to match the rest of the arguments. A bit hard to keep it short, but something like ''Filter files with the default exclusions (files must be actual paths not globs)". 

However I was thinking: why not use `files = list_files(args.files, EXCLUSIONS, TOP)` instead of listing files twice then filtering? That would still work with yourintended usage but also allow globs, making it much more useful, and no need to call list_files twice.

projectgus replied:
> However I was thinking: why not use files = list_files(args.files, EXCLUSIONS, TOP)

That doesn't work, at least for the pre-commit use case, because there is an inclusions list as well (`PATHS` variable). pre-commit needs to only check files which are simultaneously included in `PATHS` and not excluded by `EXCLUSIONS`. But you're right it would be possible to call `list_files` twice and support globbing here, I've done that and tried to make the logic as concise as possible.

> For consistency I'd reword that to match the rest of the arguments.

I've shortened it a bit. It's still a bit longer than the others, but that's because there is actually some subtlety in the behaviour here and I think it's better to try and explain that to the user.

stinos commented on tools/codeformat.py:
```diff
@@ -161,6 +166,16 @@ def main():
+            ]
+            for r in args.files:
+                if args.v and r not in filtered:
+                    print("Not checking {}".format(r))
```
Not sure it's worth making this verbose, the rest of the tool is also silent and excludes files without saying so. But maybe there's a reason you did this?

projectgus replied:
If you pass `-v` then uncrustify and black will each print a line for each file they check, even though nothing is printed by the Python script.

If this line is printed, it means the user passed a specific file name on the command line but the tool decided to ignore that file. IMO it's good to have a way to show the user that this is happening, in the spirit of "principle of least surprise".

stinos replied:
Sorry again, I missed the `if arg.v` part :)

projectgus replied:
No worries, thanks for looking over the PR! :+1:

stinos commented on tools/verifygitlog.py:
```diff
@@ -76,21 +81,23 @@ def warning(err):
-            print("warning:", err)
+
+    if "--check-message" in args:
+        has_errors, has_warnings = run_check_message(args[-1])
```
if you instead make this `errors, warnings = ...` then the code below can be reused instead of duplicating?

projectgus replied:
Not without a few more changes, because the other part runs inside a `for` loop across multiple commits.

However, you're right that this code is a bit awkward. I tried to minimise changes to the other parts, but as a result it's "smelly". I'll do a bit more refactoring.

dpgeorge replied:
I agree with @stinos.  Would be good to have only one location where there is `def error(...) ...` and `def warning(...) ...` and `for err in errors: print(...)`.  The messages come either from a file or from a git sha, but after that the code should be the same to process the message and print errors/warnings.

dpgeorge replied:
Oh, I see you refactored in the following commit...

dpgeorge replied:
The refactor is fine.  There are better things to spend time on than making this script perfect :smile:

stinos commented:
Ok thanks for the explanation!

projectgus committed:
tools/verifygitlog: Refactor error tracking.

Previous commit made this is a bit fiddly and copy-pastey, try to clean
that up.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@jimmo I found a gotcha with rebasing and verifygitlog.py, had to add logic to skip checking autosquash patterns... Rebasing it working smoothly for me now though, although if more of these show up then it might be easier to patch "gitlint" or write a plugin for what MP needs.

stinos commented on tools/codeformat.py:
```diff
@@ -161,6 +168,16 @@ def main():
+                f for f in args.files if os.path.normpath(os.path.join(TOP, f)) in all_files
+            ]
+            for r in args.files:
+                if args.v and r not in filtered:
```
Makes a bit more sense to never enter this loop if not needed, i.e. `args.v` is loop-invariant so use
```
if args.v:
   for ....
```

projectgus replied:
Good idea, at least for readability (the list of files is not long enough for it to matter for performance). Changed.

dpgeorge commented on tools/verifygitlog.py:
```diff
@@ -100,11 +113,40 @@ def run(args):
     print("-n  : do not print multi-line suggestions")
     print("-h  : print this help message and exit")
+    print(
+        "--check-message : Pass a single argument which is a file containing a candidate commit message"
```
I think this should be `--check-file` to indicate it reads from a file rather than git sha's.  (And because this tool always checks messages, regardless of where they come from.)

Also, feel free to switch to use `argparse`, but only if it's easier, don't want to spend too much time on this simple script!

projectgus replied:
@dpgeorge Thanks for the review. Done, and rebased!

I had a quick look at argparse, but it's a bit tricky - not least because it supports passing arbitrary command line options through to `git log`. I'm sure it's possible, but I'm going to opt to kick that can down the road until next time a feature needs to be added here. :)

dpgeorge commented:
Thanks for updating.  Merged the first commit as-is in bdac8272d8b1a726012bec897dc51dacbe70b295, the other 2 were squashed and merged in 0e35c4de9b9d55f9288eeb908efd2f7f577dc13b
